Add a timestamp for anti-replay (#88)
ietf-wg-ohai/oblivious-http: draft-ietf-ohai-ohttp.md

# HPKE Encapsulation

HTTP message encapsulation uses HPKE for request and response encryption.

An encapsulated HTTP request includes a timestamp and a binary-encoded HTTP
message {{BINARY}}; see {{fig-req-pt}}.

~~~
Request {

>   Timestamp (32),

Reviewer: 32 bits of seconds seems quite wide, though it is the smallest reasonable power of 2 that buys us at least a day's worth of clock skew. In a perfect world, I'd prefer this be 16 bits or shorter. Do we have up-to-date data on clock skew we can use to make a more informed choice here?

Reviewer: cc @tfpauly and @davidben, who might have a better view from the client

risk, though it is important to recognize that different idempotent requests
can be combined to be not idempotent.

Even without strong replay prevention, the server-chosen `response_nonce` field
ensures that responses have unique AEAD keys and nonces even when requests are
replayed.

### Use of Timestamps for Anti-Replay

Request messages include a timestamp from the client.  This value allows a
server to limit the amount of requests it needs to track when preventing
replays.

A server can maintain state for requests for a small window of time over which
it wishes to accept requests.  The server then rejects requests if the request

> is the same as one that was previously answered within that time window, which
> might use the `enc` value.  The server also rejects requests if the timestamp is

Author: ```suggestion
is the same as one that was previously answered within that time window. Servers
can identify duplicate requests using the `enc` value.  The server also rejects requests if the timestamp is
```

> outside of the chosen time window.  For real-time operation, servers should

Author: ```suggestion
outside of the chosen time window.  Servers SHOULD
```

allow for the time it takes requests to arrive from the client, with a time
window that is large enough to allow for differences between client and server clocks.

How large a time window is needed could depend on the population of clients that
the server needs to serve.  Unlike anti-replay protections in TLS (see {{Section
8 of TLS}}), which correct for differences in the absolute value of client and
server clocks, no system is provided to automatically account for differences in
the absolute value of client and server clocks.  This means that servers might
might need to accept and track requests over larger time windows than would be

> used for TLS. Indeed, in practice, client clock skew can be as large as a day

Author: ```suggestion
used for TLS. In practice, client clock skew can be as large as a day
```
